### ISSUE TITLE
Toggle VoteExtensions using a consensus parameter

### ISSUE THREAD

**Issue by williambanfield** [tendermint/tendermint#8453]
As of v0.36 VoteExtensions are set to be required by Tendermint. This poses a challenge to validators wishing to upgrade to v0.36 without performing a hardfork upgrade.

**Comment by thanethomson** [comment]
This makes sense to me. I like the idea of specifying a height instead of using a boolean value - it definitely seems clearer and more robust.

**Comment by thanethomson** [comment]
Just so I understand fully, looking at [the `ConsensusParams` struct](https://github.com/tendermint/tendermint/blob/37287ead94aa010f2497a5df414c64b85e4861ce/types/params.go#L39), is your idea to make something like the following change?

```golang
type ConsensusParams struct {
    VoteExtensions VoteExtensionsParams
    // ... existing fields
}

type VoteExtensionsParams struct {
    InitialHeight int64
}
```

If so, do you anticipate that we'd potentially add more vote extensions-related consensus parameters later? If not I'd advocate for something a little more generic, like `ABCIParams` with one field so far: `VoteExtensionsHeight`.

**Comment by sergio-mena** [comment]
Thanks @williambanfield for this proposal. It seems sound to me.
If folks don't have further concerns on #8317 (I have addressed all outstanding comments), I propose the following as next steps:
1) #8317 (RFC  017, v1) gets approved and merged
2) @williambanfield (or myself, if you're still not recovered) turns the description of this issue into additional paragrahps/sections of RFC017
3) The new version of RFC017 should explain how the different solutions explained should be adapted to remain correct, in particular (longer-term) solution 3. I could take this on
4) Implementation done on top of #8433 (or once #8433 is merged)

Steps 3) and 4) can be carried out in parallel.

Let me know what you think

**Comment by sergio-mena** [comment]
One last thought: Should Tendermint "automatically" set `VoteExtensions.InitialHeight` to 1 in the case we are starting the chain from scratch in v0.36 or later?
Maybe we can set a genesis param, with the same semantics as `VoteExtensions.InitialHeight`, but with default value set to 1. So chains migrating to v0.36 are not affected by the genesis param (so default value: 0). And chains starting in v0.36 will take the genesis param (default value: 1), _unless_ the App explicitly provides another value (e.g., 0).

**Comment by creachadair** [comment]
> One last thought: Should Tendermint "automatically" set `VoteExtensions.InitialHeight` to 1 in the case we are starting the chain from scratch in v0.36 or later?

We could certainly do that, but given some of the subtleties associated with using vote extensions my intuition is we might be better off making this an explicit opt-in even for new chains.

**Comment by sergio-mena** [comment]
> We could certainly do that, but given some of the subtleties associated with using vote extensions my intuition is we might be better off making this an explicit opt-in even for new chains.

Both ways seem reasonable to me. No strong opinion on either at this time.

**Comment by williambanfield** [comment]
@thanethomson 

> Just so I understand fully, looking at [the ConsensusParams struct](https://github.com/tendermint/tendermint/blob/37287ead94aa010f2497a5df414c64b85e4861ce/types/params.go#L39), is your idea to make something like the following change...

Either name is probably fine. I can't predict if we'll have more vote extension parameters so it's hard to say. My only qualm with the `ABCIParams` name is that vote extensions are not strictly an ABCI feature. Accepting / rejecting empty vote extensions has impacts on consensus underneath the conceptual ABCI barrier.

**Comment by williambanfield** [comment]
@sergio-mena I'm happy to take on the steps outlined above, although I would strongly prefer to wait until the work is completed in those PRs. I find it confusing to review PRs that expand in scope and change while they are open since you need to simultaneously track what all of the changes the PR aims to make are and what changes are actually made to match that. For me, that usually becomes too expansive to fit into my head, so I'd prefer to make the changes in serial.

**Comment by thanethomson** [comment]
> My only qualm with the ABCIParams name is that vote extensions are not strictly an ABCI feature. 

I was thinking more of it as ABCI(++)-related consensus params, since they'd be embedded in the `ConsensusParams` struct.

**Comment by thanethomson** [comment]
> However, from the use cases that have been proposed thus far, vote extensions are a pivotal piece of applications. For applications taking advantage of threshold decryption and oracle data for pricing, this is a pretty central piece of their design and execution. If they disable the feature, they will also need to fundamentally alter their application. It therefore seems safe to make vote extensions required after being enabled.

So in this regard, what if application developers want to roll back enabling vote extensions, or make them optional within the application itself? I suppose as a fallback the application itself could just ignore and not generate vote extensions if it wants, and we'd continue to just sign empty vote extensions once they're enabled.

**Comment by sergio-mena** [comment]
>So in this regard, what if application developers want to roll back enabling vote extensions, or make them optional within the application itself? I suppose as a fallback the application itself could just ignore and not generate vote extensions if it wants, and we'd continue to just sign empty vote extensions once they're enabled.

Yes, this the way I see it: once extensions are activated, the app can choose to ignore the feature altogether by always extending with empty byte arrays. And that's what it'll see at `PrepareProposal` time: empty extensions
